src/jobservice/job/impl/gc/garbage_collection.go

```diff
@@ -186,6 +114,162 @@ func (gc *GarbageCollector) init(ctx job.Context, params job.Parameters) error {
+		return nil
+	}
+
+	// get gc candidates, and set the repositories.
+	// AS the reference count is calculated by joining table project_blob and blob, here needs to call removeUntaggedBlobs to remove these non-used blobs from table project_blob firstly.
+	if !gc.dryRun {
+		gc.removeUntaggedBlobs(ctx)
+	}
+	blobs, err := gc.blobMgr.UselessBlobs(ctx.SystemContext(), gc.timeWindowHours)
+	if err != nil {
+		gc.logger.Errorf("failed to get gc candidate: %v", err)
+		return err
+	}
+	if len(blobs) == 0 {
+		gc.logger.Info("no need to execute GC as there is no non referenced artifacts.")
+		return nil
+	}
+
+	// update delete status for the candidates.
+	blobCt := 0
+	mfCt := 0
+	for _, blob := range blobs {
+		if !gc.dryRun {
+			blob.Status = blob_models.StatusDelete
+			_, err := gc.blobMgr.UpdateBlobStatus(ctx.SystemContext(), blob)
+			if err != nil {
+				gc.logger.Errorf("failed to mark gc candidate, skip it.: %s", blob.Digest)
+				continue
+			}
+		}
+		gc.logger.Infof("blob eligible for deletion: %s", blob.Digest)
+		gc.deleteSet = append(gc.deleteSet, blob)
+		// as table blob has no repository name, here needs to use the ArtifactTrash to fill it in.
+		if blob.IsManifest() {
+			for _, at := range arts {
+				if at.Digest == blob.Digest {
+					blob.Repositories = append(blob.Repositories, at.RepositoryName)
+				}
+			}
+			mfCt++
+		} else {
+			blobCt++
+		}
+	}
+	gc.logger.Infof("%d blobs and %d manifests eligible for deletion", blobCt, mfCt)
+	return nil
+}
+
+func (gc *GarbageCollector) sweep(ctx job.Context) error {
+	gc.logger = ctx.GetLogger()
+	for _, blob := range gc.deleteSet {
+		// set the status firstly, if the blob is updated by any HEAD/PUT request, it should be fail and skip.
+		blob.Status = blob_models.StatusDeleting
+		_, err := gc.blobMgr.UpdateBlobStatus(ctx.SystemContext(), blob)
+		if err != nil {
+			gc.logger.Errorf("failed to mark gc candidate deleting, skip: %s, %s", blob.Digest, blob.Status)
+			continue
+		}
+
+		if blob.IsManifest() {
+			for _, repo := range blob.Repositories {
+				// Harbor cannot know the existing tags in the backend from its database, so let the v2 DELETE manifest to remove all of them.
+				gc.logger.Infof("delete the manifest with registry v2 API: %s, %s, %s",
+					repo, blob.ContentType, blob.Digest)
+				if err := v2DeleteManifest(repo, blob.Digest); err != nil {
+					gc.logger.Errorf("failed to delete manifest with v2 API, %s, %s, %v", repo, blob.Digest, err)
+					if err := gc.markDeleteFailed(ctx, blob); err != nil {
+						return err
+					}
+					return errors.Wrapf(err, "failed to delete manifest with v2 API: %s, %s", repo, blob.Digest)
+				}
+				// for manifest, it has to delete the revisions folder of each repository
+				gc.logger.Infof("delete manifest from storage: %s", blob.Digest)
+				if err := ignoreNotFound(func() error {
+					return gc.registryCtlClient.DeleteManifest(repo, blob.Digest)
+				}); err != nil {
```

Comment:
I think the following is more clear than the wrap of the function.
```go
if err := gc.registryCtlClient.DeleteManifest(repo, blob.Digest); ignoreNotFound(err) != nil {
    // ....
}
```

Reply:
@reasonerjt your idea, both are ok for me.

Reply:
@reasonerjt your idea, both are ok for me.

Reply:
both are ok IMO